src/core/midi/MidiJack.cpp

```diff
@@ -0,0 +1,243 @@
+static int JackMidiProcessCallback(jack_nframes_t nframes, void *arg)
+{
+	MidiJack *jmd = (MidiJack *)arg;
+
+	if (nframes <= 0)
+		return (0);
+
+	jmd->JackMidiRead(nframes);
+	jmd->JackMidiWrite(nframes);
+
+	return (0);
+}
+
+static void JackMidiShutdown(void *arg)
+{
+	// TODO: support translations here
+	const QString mess_short = "JACK server down";
+	const QString mess_long = "The JACK server seems to have been shutdown.";
+	QMessageBox::information( gui->mainWindow(), mess_short, mess_long );
+}
+
+MidiJack::MidiJack() :
+	MidiClientRaw(),
+	m_input_port( NULL ),
+	m_output_port( NULL ),
+	m_quit( false )
+{
+	// if jack is used for audio then we share the connection
+	// AudioJack creates and maintains the jack connection
+	// and also handles the callback, we pass it our address
+	// so that we can also process during the callback
+
+	if(Engine::mixer()->audioDevName() == AudioJack::name() )
+	{
+		// if a jack connection has been created for audio we use that
+		m_jackAudio = dynamic_cast<AudioJack*>(Engine::mixer()->audioDev())->addMidiClient(this);
+	}else{
+		m_jackAudio = NULL;
+		m_jackClient = jack_client_open(probeDevice().toLatin1().data(),
+										JackNoStartServer, NULL);
+
+		if(m_jackClient)
+		{
+			jack_set_process_callback(m_jackClient,
+							JackMidiProcessCallback, this);
+			jack_on_shutdown(m_jackClient,
+							JackMidiShutdown, 0);
+		}
+	}
+
+	if(jackClient())
+	{
+		m_output_port = jack_port_register(
+				jackClient(), "midi_TX", JACK_DEFAULT_MIDI_TYPE,
+				JackPortIsOutput, 0);
+
+		m_input_port = jack_port_register(
+				jackClient(), "midi_RX", JACK_DEFAULT_MIDI_TYPE,
+				JackPortIsInput, 0);
+
+		if(jack_activate(jackClient()) == 0 )
+		{
+			// only start thread, if we have an active jack client.
+			start( QThread::LowPriority );
+		}
+	}
+}
+
+MidiJack::~MidiJack()
+{
+	if(jackClient())
+	{
+		if( jack_port_unregister( jackClient(), m_input_port) != 0){
+			printf("Failed to unregister jack midi input\n");
+		}
+
+		if( jack_port_unregister( jackClient(), m_output_port) != 0){
+			printf("Failed to unregister jack midi output\n");
+		}
+
+		if(m_jackClient)
+		{
+			// an m_jackClient means we are handling the jack connection
+			if( jack_deactivate(m_jackClient) != 0){
+				printf("Failed to deactivate jack midi client\n");
+			}
+
+			if( jack_client_close(m_jackClient) != 0){
+				printf("Failed close jack midi client\n");
+			}
+		}
+	}
+	if( isRunning() )
+	{
+		m_quit = true;
+		wait( 1000 );
+		terminate();
+	}
+}
+
+jack_client_t* MidiJack::jackClient()
+{
+	if( m_jackAudio == NULL && m_jackClient == NULL)
+		return NULL;
+
+	if( m_jackAudio == NULL && m_jackClient )
+		return m_jackClient;
+
+	return m_jackAudio->jackClient();
+}
+
+QString MidiJack::probeDevice()
+{
+	QString jid = ConfigManager::inst()->value( "midijack", "lmms" );
+	if( jid.isEmpty() )
+	{
+		return "lmms";
+	}
+	return jid;
+}
+
+void MidiJack::sendByte( const unsigned char c )
+{
+	//m_midiDev.putChar( c );
```

Comment:
Unimplemented - I haven't found how to make it work (and not 100% certain how it is meant to work to test if it is right). It exists in the other midi clients, I assume it is needed to get midi out to work and it gets called from MidiClientRaw::processOutEvent. The putChar there is from MidiOss.

MidiJack::JackMidiWrite is used during the jack callback event,  I expect MidiJack::sendByte should write to a queue from the main thread that can then be safely read later during the jack callback event.